### ISSUE TITLE
bad signature when attaching file with name longer than 34 chars

### ISSUE THREAD

**Issue by quite** [CZ-NIC/envelope#19]
~~I'm trying to attach a text/plain file to my envelope that I also want signature() for. I get a bad signature unless I also pass a `name=` for the file.~~ Continue reading pls :)

**Comment by e3rd** [comment]
Please provide more information.

**Comment by quite** [comment]
Sorry. I've tried a bit more. Somewhat confusingly, I get a bad signature when an attachment has a name longer than 34 chars. I'm verifying using mutt, don't know any other way currently. Chaing `n` to 35 below reproduces the error for me. I have envelope==1.5.3

```
 n = 34

 e = (
     Envelope()
     .from_("me@example.com")
     .to("me@example.com")
     .subject("subject")
     .message("body")
     .attach("some data", name="X" * n)
     .signature()
 )

 e = e.send()
```

**Comment by e3rd** [comment]
Cool, thanks! I ll try to investigate next week!

**Comment by e3rd** [comment]
Unfortunately, the code you provided seems to not be able to reproduce the error for me. Append the error report too, please.

**Comment by quite** [comment]
I've now managed to verify the signatures using python, code below. And the all verify, no matter attachment filename. It looks like I've run into a bug in mutt?!

```
import email
import os
for mailfn in sys.argv[1:]:
  with open(mailfn) as f:
    msg = email.message_from_file(f)
    with open("txt", "wb") as textf:
      textf.write(msg.get_payload(0).as_bytes())
    with open("sig", "wb") as sigf:
      sigf.write(msg.get_payload(1).as_bytes())
    os.system("gpg --verify sig txt")
```

**Comment by e3rd** [comment]
So there is not error log in the terminal while sending it but the e-mail client mutt tells "BAD signature" and won't display the text? I'd really like to get the e-mail source code produced by the library, either here publicly or somehow privately. Are you able to get me the source code? (Ctrl+U in Thunderbird to display e-mail source.)

**Comment by quite** [comment]
Yes, mailed you at github@

**Comment by e3rd** [comment]
Thanks! I received and I'll try to investigate :)

**Comment by quite** [comment]
I filed an issue with mutt here: https://gitlab.com/muttmua/mutt/-/issues/382

Also I just tried having Thunderbird verify the emails. It fails on the n=35 one as well! Hmmmm...

**Comment by quite** [comment]
I've generated a dummy gpgkey and recreated the signed mails using it. Maildir, GPGHOME and python scripts here: 
[signtest.tar.gz](https://github.com/CZ-NIC/envelope/files/7815714/signtest.tar.gz)

**Comment by quite** [comment]
@e3rd There might be two issues. See https://gitlab.com/muttmua/mutt/-/issues/382#note_802356286

1) signature is made over a wrapped header-line, namely `Content-Disposition`. I seems to be that headers must not be wrapped when making the signature. My verification code above is then doing the same mistake, as `get_payload()` apparently returns it wrapped. (I don't know if that is bad in itself)

2) I seems that the signature also must be made over text which uses the CR-LF line ending convention. Envelope might already be doing this?

**Comment by quite** [comment]
Update: perhaps it doesn't matter if the headers are folded or not. But what matters is that the text which is sign is exactly the text that is written to the mail. And that does not seem to be the case when a long attachment filename ends up causing the header to wrap. The signature is made over the text where the header is folded, but in the final mail that Envelope generates, it is not folded.

**Comment by quite** [comment]
in `_compose_gpg_signed()`, the `email` parameter is what has been signed (and `text` is the actual signature). Here I can print `str(email)` and see that the content-disposition header has been folded (when n=35).

Later, after `email = email.attach(msg_payload)`, printing `str(email)` gives that header no longer folded. This is what ends up in the actual email.

I'm on python 3.10.1 btw.

**Comment by e3rd** [comment]
(thanks for investigating, append .eml source every time please so that it is clearly visible for me, I had not have the time to dive into this yet)

**Comment by quite** [comment]
Maybe this "policy" https://docs.python.org/3/library/email.policy.html is something which can be used to avoid the unexpected header (un)folding. Haven't read or understood it.

**Comment by e3rd** [comment]
The two e-mail you sent me on Jan 4 2022 seem valid to me. Could you please repeat what is the core of the problem (something with header folding, I know). Append a code stub that I can re-run to produce an invalid e-mail source output so that I understand what's going on and see an error. Thanks!

**Comment by quite** [comment]
I don't have much time... Please read this again carefully: https://github.com/CZ-NIC/envelope/issues/19#issuecomment-1006519784  -- there I could identify the problem. In that function, the 2 print()s that I mention shows that what has been signed (1st print) is *not* the same same as what ends up in the email (2nd print).

**Comment by e3rd** [comment]
I beg you to show the output, not just describe

**Comment by quite** [comment]
Right. I'm running the create_mails.py script (from https://github.com/CZ-NIC/envelope/issues/19#issuecomment-1005757425), but only for n=35 (the one the causes the issue).

Envelope.py patched like this:
```
--- venv-unpatched/lib/python3.10/site-packages/envelope/envelope.py    2021-06-03 11:09:12.000000000 +0200
+++ venv/lib/python3.10/site-packages/envelope/envelope.py      2022-01-21 08:50:41.239219465 +0100
@@ -1247,6 +1247,8 @@
         return output_buffer.read()

     def _compose_gpg_signed(self, email, text, micalg=None):
+        with open("./email-that-was-signed", "wb") as f:
+            f.write(email.as_bytes())
         msg_payload = email
         email = EmailMessage()
         email["Subject"] = self._subject
@@ -1255,6 +1257,8 @@
         if micalg:
             email.set_param("micalg", micalg)
         email.attach(msg_payload)
+        with open("./email-that-will-be-sent", "wb") as f:
+            f.write(email.as_bytes())
         msg_signature = EmailMessage()
         msg_signature['Content-Type'] = 'application/pgp-signature; name="signature.asc"'
         msg_signature['Content-Description'] = 'OpenPGP digital signature'
```

Contents of `email-that-was-signed`:
```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="===============5796393638397195874=="

--===============5796393638397195874==
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 7bit

body body body body

--===============5796393638397195874==
Content-Type: text/plain
Content-Transfer-Encoding: base64
Content-Disposition: attachment;
 filename="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
MIME-Version: 1.0

c29tZSBkYXRhIEhFSiBPIEhP

--===============5796393638397195874==--
```

Contents of `email-that-will-be-sent`:
```
Subject: n=35
MIME-Version: 1.0
Content-Type: multipart/signed; protocol="application/pgp-signature";
 micalg="pgp-sha256"; boundary="===============7624938521358705995=="

--===============7624938521358705995==
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="===============5796393638397195874=="

--===============5796393638397195874==
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 7bit

body body body body

--===============5796393638397195874==
Content-Type: text/plain
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
MIME-Version: 1.0

c29tZSBkYXRhIEhFSiBPIEhP

--===============5796393638397195874==--

--===============7624938521358705995==--
```

Diff between the 2 output files:
```
--- email-that-was-signed       2022-01-21 08:50:42.089221620 +0100
+++ email-that-will-be-sent     2022-01-21 08:50:42.099221644 +0100
@@ -1,3 +1,9 @@
+Subject: n=35
+MIME-Version: 1.0
+Content-Type: multipart/signed; protocol="application/pgp-signature";
+ micalg="pgp-sha256"; boundary="===============7624938521358705995=="
+
+--===============7624938521358705995==
 MIME-Version: 1.0
 Content-Type: multipart/mixed; boundary="===============5796393638397195874=="

@@ -10,10 +16,11 @@
 --===============5796393638397195874==
 Content-Type: text/plain
 Content-Transfer-Encoding: base64
-Content-Disposition: attachment;
- filename="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+Content-Disposition: attachment; filename="XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
 MIME-Version: 1.0

 c29tZSBkYXRhIEhFSiBPIEhP

 --===============5796393638397195874==--
+
+--===============7624938521358705995==--
```

As you can see, the signature is made over a folded header-line. What is sent contains the same header unfolded. My `verify_emails.py` script in the above tar-ball wrongly verifies this signature correctly, because it ends up making the same kind of error. But I'm still not sure what the error is, or *where* precisely it must be correct. But it might have to do with the email.policy that I mentioned above.

**Comment by quite** [comment]
> Append a code stub that I can re-run to produce an invalid e-mail source output so that I understand what's going on and see an error. Thanks!

Btw, did you actually run the code in this tar-ball? https://github.com/CZ-NIC/envelope/issues/19#issuecomment-1005757425
It produces two emails. One with valid and one with invalid signature, as verified by Thunderbird, mutt and also emacs/notmuch.el-setup. But as I said, even my simple Python code (in the same tar-ball)  is flawed, and successfully verifies both emails.

**Comment by e3rd** [comment]
TL;DR
Standard Python library went wild.

Longer:
Really grateful that you have posted this issue and added the tarballs etc. That was a real Sherlock trip I had yesterday. Just like the guys from mutt said... In the package email.generator, the method Generator._handle_multipart_signed states: "The contents of signed parts has to stay unmodified in order to keep the signature intact per RFC1847 2.1, so we disable header wrapping." And for an instant, it changes the policy to not wrap (I tried really hard and vain to come up with a use-case for such a strange behaviour). Any policy mangling failed because of this internal behaviour of an official Python module: signing a wrapped message does not work as it is output unwrapped and not wrapping the message is not an option either due to as it seems another bug.
So I've done some monkey so that envelope works and raised an official issue so that standard library might get better https://github.com/python/cpython/issues/89789
